File: internal/gcs/gcs.go
package gcs

import (
	"context"
	"io"
	"log"
	"os"
	"path/filepath"

	"cloud.google.com/go/storage"
	"github.com/jaluprayoga/car-price-prediction-deployment/internal/config"
	"google.golang.org/api/option"
)

// getGCSClient initializes a Google Cloud Storage client based on env configuration.
func getGCSClient(ctx context.Context) (*storage.Client, error) {
	credsPath := config.AppConfig.GoogleCredentials
	if credsPath != "" {
		// Resolve relative path based on current directory
		if !filepath.IsAbs(credsPath) {
			// Check if file exists relative to working dir
			if _, err := os.Stat(credsPath); err != nil {
				log.Printf("[GCS] Credentials file specified not found: %s, attempting default auth", credsPath)
				return storage.NewClient(ctx)
			}
		}
		log.Printf("[GCS] Initializing client using credentials from: %s", credsPath)
		return storage.NewClient(ctx, option.WithCredentialsFile(credsPath))
	}

	log.Printf("[GCS] Initializing client using default credentials")
	return storage.NewClient(ctx)
}

// DownloadFromGCS downloads a file from GCS. Returns true on success, false on failure.
func DownloadFromGCS(localFilePath, gcsBlobName, bucketName string) bool {
	if bucketName == "" {
		bucketName = config.AppConfig.GCSBucketName
	}

	ctx := context.Background()
	client, err := getGCSClient(ctx)
	if err != nil {
		log.Printf("[GCS] Client unavailable. Skipping download of '%s' from cloud storage: %v", gcsBlobName, err)
		return false
	}
	defer client.Close()

	log.Printf("[GCS] Fetching blob '%s' from bucket '%s'...", gcsBlobName, bucketName)
	bucket := client.Bucket(bucketName)
	blob := bucket.Object(gcsBlobName)

	// Open reader
	reader, err := blob.NewReader(ctx)
	if err != nil {
		log.Printf("[GCS] Failed to open reader for blob '%s': %v", gcsBlobName, err)
		return false
	}
	defer reader.Close()

	// Ensure destination directory exists
	dir := filepath.Dir(localFilePath)
	if err := os.MkdirAll(dir, 0755); err != nil {
		log.Printf("[GCS] Failed to create directory '%s': %v", dir, err)
		return false
	}

	// Open destination file
	outFile, err := os.Create(localFilePath)
	if err != nil {
		log.Printf("[GCS] Failed to create destination file '%s': %v", localFilePath, err)
		return false
	}
	defer outFile.Close()

	// Copy content
	_, err = io.Copy(outFile, reader)
	if err != nil {
		log.Printf("[GCS] Failed to write downloaded content to file: %v", err)
		return false
	}

	log.Printf("[GCS] Successfully downloaded '%s' from GCS to '%s'", gcsBlobName, localFilePath)
	return true
}

File: src/utils/gcs.py
import os
import logging
from pathlib import Path
from dotenv import load_dotenv
from google.cloud import storage

logger = logging.getLogger("deployment.gcs")

# Resolve workspace root and load .env dynamically
current_dir = Path(__file__).resolve()
env_path = None
for parent in current_dir.parents:
    if (parent / ".env").exists():
        env_path = parent / ".env"
        break

if env_path:
    load_dotenv(env_path)
    logger.info(f"Loaded environment variables from: {env_path}")
else:
    logger.warning("No .env file found in parent directories. Using environment variables directly.")

def get_gcs_client():
    """Initializes the GCS client, setting credentials from GOOGLE_APPLICATION_CREDENTIALS if provided."""
    creds_path = os.environ.get("GOOGLE_APPLICATION_CREDENTIALS")
    if creds_path:
        creds_file = Path(creds_path)
        if not creds_file.is_absolute() and env_path:
            root_relative = env_path.parent / creds_path
            if root_relative.exists():
                creds_path = str(root_relative)
        
        if os.path.exists(creds_path):
            try:
                return storage.Client.from_service_account_json(creds_path)
            except Exception as e:
                logger.warning(f"Failed to initialize GCS client from service account json at {creds_path}: {e}. Falling back to default auth.")
        else:
            logger.warning(f"Credentials file specified in GOOGLE_APPLICATION_CREDENTIALS not found: {creds_path}")
            
    try:
        return storage.Client()
    except Exception as e:
        logger.warning(f"Could not initialize GCS client: {e}. GCS operations will be bypassed (local fallback).")
        return None

def download_from_gcs(local_file_path: str, gcs_blob_name: str, bucket_name: str = None) -> bool:
    """Downloads a file from GCS. Returns True on success, False on failure (e.g. if offline or not found)."""
    if not bucket_name:
        bucket_name = os.environ.get("GCS_BUCKET_NAME", "car-price-prediction-mlops")
        
    client = get_gcs_client()
    if not client:
        logger.info(f"GCS client unavailable. Skipping download of '{gcs_blob_name}' from cloud storage.")
        return False
        
    try:
        bucket = client.bucket(bucket_name)
        blob = bucket.blob(gcs_blob_name)
        
        os.makedirs(os.path.dirname(local_file_path), exist_ok=True)
        blob.download_to_filename(local_file_path)
        logger.info(f"Successfully downloaded '{gcs_blob_name}' from GCS to '{local_file_path}'")
        return True
    except Exception as e:
        logger.error(f"Failed to download '{gcs_blob_name}' from GCS: {e}")
        return False
